extensions/amp-ad/0.1/amp-ad-image.js

```diff
@@ -0,0 +1,252 @@
+/**
+ * Copyright 2016 David Karlin, Bachtrack Ltd. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {isLayoutSizeDefined} from '../../../src/layout';
+import {getLifecycleReporter} from '../../../ads/google/a4a/performance';
+import {user} from '../../../src/log';
+
+
+
+/** @const {!string} Tag name for 3P AD implementation. */
+export const TAG_AD_IMAGE = 'amp-ad-image';
+
+/** @const {int} The number of milliseconds between checks to see if the response has arrived */
+const IMAGEAD_TICK_LENGTH = 50;
+
+/** @const {int} The number of ticks to wait for a response from the ad server before timing out */
+const IMAGEAD_TIMEOUT = 100;
+
+export class AmpAdImage extends AMP.BaseElement {
+
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /** @private {string} The base URL of the ad server for this ad */
+    this.url_ = element.getAttribute('data-url');
+
+    /** @private {string} A string identifying this ad slot: the server's responses will be keyed by slot */
+    this.slot_ = element.getAttribute('data-slot');
+
+    /** @private {string} Target specifier for the ad's anchor tag: default to "_blank", accept "_top" */
+    this.target_ = '_blank';
+    const t = element.getAttribute('data-target');
+    if (t) {
+      if (t === '_blank' || t === '_self') {
+        this.target_ = t;
+      } else {
+        user().error(TAG_AD_IMAGE,
+        "Invalid data-target: only '_blank' and '_self' are permitted");
+      }
+    }
+
+
+    // If this is our first imagead, create a map of responses for each value of 'data-url'
```

Comment:
A few comments about the implementation:

1. can you create a self-contained class for that purpose? something like `CustomAdBatchingManager`.

2. I noticed you register each slot in `buildCallback` and send the request in the first `layoutCallback`. There's risk that not all slots get registered before we send the request. An alternative can be composing the request URL in the BatchingManager, when it receives the first buildCallback from any slot. It scan the doc, and find all info it needs to build request URLs. (it's a bit dirty, but the best way I can think of).